### PR TITLE
Add ability to modify provider options at runtime

### PR DESCRIPTION
**Pull request by RyanUnderhill**
Now a provider can be chosen and configured purely at runtime, to override what the genai_config.json specifies.
`
  auto config = OgaConfig::Create("path\to\model");
  config->SetProviderOption("cuda", "enable_quantum_drive", "1");
  auto model = OgaModel::Create(*config);
`
vs this if no runtime changes are needed (this is also the current API):
`
  auto model = OgaModel::Create("path\to\model");
`

This will solve the following use cases:
* User wants to change a runtime option for a provider already specified in the genai_config.json
* User wants to switch which provider is used vs what the genai_config.json specifies
* The genai_config.json specifies no provider, and one is completely chosen at runtime.

The OgaConfig provides these methods:
* ClearProviders() - Remove all providers currently specified (will default to cpu if nothing else happens)
* SetProvider("provider_name") - If provider is already exists, does nothing, otherwise adds to end of provider list
* SetProviderOption("provider_name", "option_name", "option_value") - If provider doesn't exist, adds it to end of list. Then sets an option on the provider.

### CODE REVIEW

File: src/java/src/main/native/ai_onnxruntime_genai_Config.cpp [hunk 1]
```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+#include "ai_onnxruntime_genai_Model.h"
```
Comment (skyline75489) [inline src/java/src/main/native/ai_onnxruntime_genai_Config.cpp:5]: This should be `ai_onnxruntime_genai_Config.h"

Comment (RyanUnderhill) [reply src/java/src/main/native/ai_onnxruntime_genai_Config.cpp]: I just noticed that too.

File: src/ort_genai_c.h [hunk 2]
```diff
@@ -172,14 +173,56 @@ OGA_EXPORT void OGA_API_CALL OgaDestroyRuntimeSettings(OgaRuntimeSettings* setti
...
+ */
+OGA_EXPORT OgaResult* OGA_API_CALL OgaConfigClearProviders(OgaConfig* config);
+
+/*
+ * \brief Add the provider at the end of the list of providers in the given config if it doesn't already exist
+ * \param[in] config The config to set the provider on.
+ * \param[in] provider The provider to set on the config.
+ * \return OgaResult containing the error message if the setting of the provider failed.
+ */
+OGA_EXPORT OgaResult* OGA_API_CALL OgaConfigSetProvider(OgaConfig* config, const char* provider);
```
Comment (pranavsharma) [inline src/ort_genai_c.h:196]: 1. So, this is the one that applies to all models in the config (if there are more than 1 model)?
2. I suppose you'll add the model specific set provider later?

File: src/ort_genai_c.h [hunk 3]
```diff
@@ -172,14 +173,56 @@ OGA_EXPORT void OGA_API_CALL OgaDestroyRuntimeSettings(OgaRuntimeSettings* setti
...
+ * \brief Creates a model from the given configuration directory.
  * \param[in] config_path The path to the model configuration directory. The path is expected to be encoded in UTF-8.
  * \param[in] device_type The device type to use for the model.
  * \param[out] out The created model.
  * \return OgaResult containing the error message if the model creation failed.
  */
 OGA_EXPORT OgaResult* OGA_API_CALL OgaCreateModel(const char* config_path, OgaModel** out);
 
+/*
+ * \brief Creates a model from the given configuration directory and device type.
```
Comment (pranavsharma) [inline src/ort_genai_c.h:218]: Documentation needs revision.

Comment (RyanUnderhill) [reply src/ort_genai_c.h:218]: Good catch, I caught and removed the 'and device type.' from the existing OgaCreateModel function but didn't realize I copied it here first.

File: src/ort_genai_c.h (hunk 2 above)
Comment (pranavsharma) [reply src/ort_genai_c.h:196]: Documentation says that you are appending, but API says Set. What is the API doing?

File: src/python/python.cpp [hunk 4]
```diff
@@ -5,6 +5,7 @@
 #include <pybind11/numpy.h>
 #include <iostream>
 #include "../generators.h"
+#include "../ort_genai.h"
```
[the remark this answers is not in the document]
Comment (RyanUnderhill) [reply src/python/python.cpp:8]: They do, the config object in python goes through the C API. I did it as a first test to start switching away from an invasive python bindings.

Python should use the same API that all of the other language bindings use.

File: src/models/debugging.cpp [hunk 5]
```diff
@@ -19,7 +19,7 @@ static constexpr size_t c_value_count = 10;  // Dump this many values from the s
 template <typename... Types>
 const char* TypeToString(ONNXTensorElementDataType type, Ort::TypeList<Types...>) {
   const char* name = "(please add type to list)";
-  ((type == Ort::TypeToTensorType<Types> ? name = typeid(Types).name(), true : false) || ...);
+  (void)((type == Ort::TypeToTensorType<Types> ? name = typeid(Types).name(), true : false) || ...);
```
[the remark this answers is not in the document]
Comment (RyanUnderhill) [reply src/models/debugging.cpp:22]: To silence a compiler warning on Android that warns of an expression with an unused result.

File: src/python/python.cpp [hunk 6]
```diff
@@ -433,7 +434,18 @@ PYBIND11_MODULE(onnxruntime_genai, m) {
...
 
+  // Pybind class using the C API for OgaConfig
+  pybind11::class_<OgaConfig>(m, "Config")
+      .def(pybind11::init([](const std::string& config_path) { return OgaConfig::Create(config_path.c_str()); }))
+      .def("set_provider", &OgaConfig::SetProvider)
+      .def("set_provider_option", &OgaConfig::SetProviderOption)
+      .def("clear_providers", &OgaConfig::ClearProviders);
+
   pybind11::class_<Model, std::shared_ptr<Model>>(m, "Model")
+      .def(pybind11::init([](const OgaConfig& config) {
```
[the remark this answers is not in the document]
Comment (RyanUnderhill) [reply src/python/python.cpp:445]: See the #include "../ort_genai.h" comment below

File: test/c_api_tests.cpp [hunk 7]
```diff
@@ -18,9 +18,24 @@
 #define PHI2_PATH MODEL_PATH "phi-2/int4/cpu"
 #endif
 #endif
+TEST(CAPITests, Config) {
+#if TEST_PHI2
```
[the remark this answers is not in the document]
Comment (RyanUnderhill) [reply test/c_api_tests.cpp:22]: Not really. But we want to remove the gpt2 models or at least convert them to something new (they're the only ones using the KV_Cache_Combined stuff.

File: test/python/test_onnxruntime_genai_api.py [hunk 8]
```diff
@@ -45,7 +54,8 @@
 def test_greedy_search(test_data_path, relative_model_path):
     model_path = os.fspath(Path(test_data_path) / relative_model_path)
 
-    model = og.Model(model_path)
+    config = og.Config(model_path) # Test using config vs path directly
+    model = og.Model(config)
```
[the remark this answers is not in the document]
Comment (RyanUnderhill) [reply test/python/test_onnxruntime_genai_api.py:58]: It only needs to be done once somewhere, I don't want to double the test run time for that.

File: test/python/test_onnxruntime_genai_api.py [hunk 9]
```diff
@@ -29,6 +29,15 @@
 if og.is_rocm_available():
     devices.append("rocm")
 
+def config_test(test_data_path):
```
[the remark this answers is not in the document]
Comment (RyanUnderhill) [reply test/python/test_onnxruntime_genai_api.py:32]: Hmm, good point

**Review by natke — COMMENTED** [review]
Can you add a code sample in the PR for the scenario that is driving this change ie to set the default provider? Did we lose the SetDefaultProvider() API?

**Comment by RyanUnderhill** [comment]
> Can you add a code sample in the PR for the scenario that is driving this change ie to set the default provider? Did we lose the SetDefaultProvider() API?

The initial 'SetDefaultProvider' proposed API would rearrange the providers so the one you mention is first in the list. This isn't really what we want since it doesn't match the onnxruntime API for setting providers.

Where should the sample go? This is how it will look:

`auto config = OgaConfig::Create("path\to\model");
config.ClearProviders(); // Only needed if config already specifies a provider, but safe to do in both cases
config.SetProvider("dml");
auto model = OgaModel::Create(*config);
`

File: src/ort_genai_c.h (hunk 2 above)
Comment (RyanUnderhill) [reply src/ort_genai_c.h:196]: It's both. If it's already in the list, it does nothing. If not, it appends it to the end.

Should I rename it to AppendProvider and have it be an error if it already exists?

Comment (pranavsharma) [reply src/ort_genai_c.h:196]: Discussed offline. Let's call it Append.

**Review by pranavsharma — APPROVED** [review]
Approving for the C/C++ API changes.

**Pull request merged**